### PR TITLE
Detach SearchPhases from AbstractSearchAsyncAction

### PR DESCRIPTION
**Pull request by s1monw**
Today all search phases are inner classes of AbstractSearchAsyncAction or one of it's
subclasses. This makes unittesting of these classes practically impossible. This commit
Extracts `DfsQueryPhase` and `FetchSearchPhase` or of the code that composes the actual
query execution types and moves most of the fan-out and collect code into an `InitialSearchPhase`
class that can be used to build initial search phases (phases that retry on shards). This will
make modification to these classes simpler and allows to easily compose or add new search phases
down the road if additional roundtrips are required.

### CODE REVIEW

**Comment by s1monw** [comment]
@elasticmachine would you bother to test this

File: core/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    private final Logger logger;
+    private final int expectedTotalOps;
+    private final AtomicInteger totalOps = new AtomicInteger();
+
+    InitialSearchPhase(String name, SearchRequest request, GroupShardsIterator shardsIts, Logger logger) {
+        super(name);
+        this.request = request;
+        this.shardsIts = shardsIts;
+        this.logger = logger;
+        // we need to add 1 for non active partition, since we count it in the total!
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java]: What do you mean by the `non-active partition` here?

**Review by colings86 — APPROVED** [review]
LGTM, but I think  a second pair of eyes form @jpountz or @jimczi would be good here too

**Comment by s1monw** [comment]
@colings86 I pushed new commits to address your comments.

File: core/src/main/java/org/elasticsearch/action/search/SearchDfsQueryThenFetchAsyncAction.java [hunk 2]
```diff
@@ -20,108 +20,42 @@
...
-                            // the query might not have been executed at all (for example because thread pool rejected
-                            // execution) and the search context that was created in dfs phase might not be released.
-                            // release it again to be in the safe side
-                            sendReleaseSearchContext(querySearchRequest.id(), connection);
-                        }
-                    }
-                });
-            }
-        }
+    protected SearchPhase getNextPhase(AtomicArray<DfsSearchResult> results, SearchPhaseContext context) {
```
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/action/search/SearchDfsQueryThenFetchAsyncAction.java]: beautiful !

**Review by jimczi — APPROVED** [review]
This is a great change that introduces unit tests for each phase and simplifies things in a complicated way ;) !
The abstraction for search phase and the chaining are just great and you preserved small optims like query and fetch for single shard query.
Looks fantastic to me.

File: core/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+                     SearchPhaseContext context, Function<SearchResponse, SearchPhase> nextPhaseFactory) {
+        super("fetch");
+        this.fetchResults = new AtomicArray<>(queryResults.length());
+        this.searchPhaseController = searchPhaseController;
+        this.queryResults = queryResults;
+        this.nextPhaseFactory =  nextPhaseFactory;
+        this.context = context;
+        this.logger = context.getLogger();
+        assert context.getNumShards() == queryResults.length() : "number of shards must match the length of the query results but doesn't:"
+             + context.getNumShards() + "!=" + queryResults.length();
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java]: let's make it a hard exception too?

**Review by jpountz — APPROVED** [review]
LGTM. It is hard to do thorough reviews for such changes, but I like where this is going. I left some nitpicks about moving assertions in non performance-sensitive code to hard exceptions.

**Comment by s1monw** [comment]
@jpountz @jimczi thanks for the feedback I pushed fixes for the last suggestions and will merge this once it's build.

**Comment by s1monw** [comment]
@elasticmachine would you bother to test this

**Comment by s1monw** [comment]
@elasticmachine would you bother to test this

**Pull request merged**